solver/src/naive_amm_settlement.rs

```diff
@@ -0,0 +1,307 @@
+    Settlement {
+        clearing_prices: maplit::hashmap! {
+            sell_a.sell_token => sell_a.sell_amount,
+            sell_b.sell_token => sell_b.sell_amount,
+        },
+        trades: vec![
+            Trade {
+                order: *sell_a,
+                executed_amount: sell_a.sell_amount,
+            },
+            Trade {
+                order: *sell_b,
+                executed_amount: sell_b.sell_amount,
+            },
+        ],
+        ..Default::default()
+    }
+}
+
+// Match two orders with amm assuming that there is price overlap.
+fn amm_match_(sell_a: &UserOrder, sell_b: &UserOrder) -> Option<Settlement> {
+    // Based on our assumptions we know that exactly one order is "bigger" than the other in the
+    // sense that it a larger sell amount than the other order's buy amount.
+    // It is not possible for both orders to be bigger because that would be a direct match which
+    // has already been handled.
+    let (big, small) = if sell_a.sell_amount > sell_b.buy_amount {
+        (sell_a, sell_b)
+    } else {
+        (sell_b, sell_a)
+    };
+    // Unwrap because of the above explanation.
+    let big_missing_buy_amount = big.buy_amount.checked_sub(small.sell_amount).unwrap();
+    // Because the smart contract enforces uniform prices we must pick one price that is accepted by
+    // both orders. We know that there is price overlap so we could pick any price between the price
+    // limits of the orders. We pick the price of the bigger order because it's sell token surplus
+    // will have to be traded with the amm and in this trade it should suffice to receive as few
+    // tokens as possible (thus using least constraining price).
+    // Because we picked the bigger order's price we must add an extra buy amount to the smaller
+    // order so that it gets the same price.
+    let small_buy_amount = rounded_up_division(
+        small.sell_amount.checked_mul(big.sell_amount)?,
+        big.buy_amount,
+    )?;
+    // Does this have to be saturating or is guaranteed to not underflow? With the rounding from the
+    // previous operation I suspect it might.
+    let big_extra_sell_amount = big.sell_amount.saturating_sub(small_buy_amount);
+    if big_extra_sell_amount.is_zero() {
+        return None;
+    }
+
+    Some(Settlement {
+        clearing_prices: maplit::hashmap! {
+            big.sell_token => big.sell_amount,
+            big.buy_token => big.buy_amount,
+        },
+        trades: vec![
+            Trade {
+                order: *sell_a,
+                executed_amount: sell_a.sell_amount,
+            },
+            Trade {
+                order: *sell_b,
+                executed_amount: sell_b.sell_amount,
+            },
+        ],
+        interactions: vec![Interaction::AmmSwapExactTokensForTokens {
+            amount_in: big_extra_sell_amount,
+            amount_out_min: big_missing_buy_amount,
```

Comment:
Are we actually doing the interaction against the uniswap pool or against the router? 

I always had in mind that we would go for the [swap contract](https://github.com/Uniswap/uniswap-v2-core/blob/master/contracts/UniswapV2Pair.sol#L159) directly, instead of the [router](https://github.com/Uniswap/uniswap-v2-periphery/blob/master/contracts/UniswapV2Router02.sol#L224)

The benefit of the router is that it calculates for us, how much we get in return. Hence `amount_out_min` makes sense. But if we want to have the efficiency of saving one "approve" for the router, then we should talk directly to the swap contract.

But yeah, I think we should use your method and go with the router. Otherwise, we might need to deploy custom price logic smart contracts. The approves are not too bad, as we can initially set them for all important tokens with unit(-1)